### PR TITLE
Update file.replace to be hardlink-friendly, fixes saltstack#20969

### LINKED ISSUE #20969: file.replace writes to a new inode, creating problems for hard links
`file.replace` relies on python's `fileinput` method, passing it the parameter `inplace=True` so that `file.replace` can write to the same filename. This parameter causes `fileinput` to *move* the target file, appending `backup` to the filename. See the [fileinput documentation][0]. Then, when <stdout> is written, `fileinput` redirects the output to a new inode that has the same filename as the target. This is a new file, because it is a new inode.

All of this happens even if `backup=False` due to the way `fileinput` works.

The original file that was moved retains it's original inode. So at this point, if there are hard links to the target file, they now point to the backup rather than original. This is likely to cause confusion for users at some point. I.e., they will edit the hard link (which they likely created in the first place), thinking it is still the same as the target file, though now its inode is shared by the backup instead.

I can think of two approaches to addressing this behavior:

1. Continue using `fileinput`, but delete any hardlinks that point to the original file and re-create them so they point to the resulting file. To do this, we'd have to get the inode of the original file, then get a list of files that point to that inode.
    - This approach is probably easiest, and the behavior could be controlled by a boolean parameter `preserve_hardlinks`.
2. Stop using `fileinput`. Instead, *copy* the target file to create the backup. By creating a copy instead of moving it, we retain the inode of the original. Then `open` the backup in 'r' mode and `open` the target in 'w' mode, and read the lines from the backup and write them to the target while searching and replacing the pattern. If `backup=False`, delete the backup file afterwards (just like `fileinput`).
    - This approach may be more 'complete' but a little more difficult to implement. If someone *really* wanted to preserve existing broken behavior, just *move* the file instead of using *copy* (with the behavior again toggled by a boolean parameter).

Other thoughts:
- I thought about using two handles to the same file, and reading from one (mode 'r') while writing to the other (mode 'r+' to avoid truncation). This would avoid the copy, but if the replacement string contains multiple lines then there are potential cases where we'd write over lines that hadn't yet been read.
- We could also avoid the copy by reading the entire file into memory and operating on that, but `file.replace` currently appears to avoid that approach intentionally. Probably to reduce resource consumption and mitigate buffer overflow attacks.

Perhaps there are other approaches as well?

> Note: I discovered this issue while researching another issue with file.replace and symlinks, reported [here][1]. Linking the two so anyone searching issues that ends up at the other one gets redirected here.

[0]: https://docs.python.org/2.7/library/fileinput.html
[1]: https://github.com/saltstack/salt/issues/20910

### PR DESCRIPTION
**Pull request by lorengordon**
Fixes #20969.

This commit modifies file.replace to use salt.utils.fopen rather fileinput
when modifying the file. This provides more control over how the file is
manipulated. Previously, fileinput would always _move_ the file, then
write to a new inode. This isn't very friendly for any hardlinks that
shared the original inode.

This commit introduces the parameter `preserve_inode` to control the
behavior of file.replace. If `True`, the file is _copied_, then
file.replace writes to the original inode. If `False`, the file is
_moved_, then file.replace writes to a new inode (mimicking prior
behavior). Default is `True`.

### CODE REVIEW

File: salt/modules/file.py [hunk 1]
```diff
@@ -1140,6 +1141,16 @@ def replace(path,
             Using this option will store two copies of the file in-memory
             (the original version and the edited version) in order to generate the
             diff.
+    preserve_inode
+        .. versionadded::
```
Comment (jfindlay) [inline salt/modules/file.py]: `.. versionadded:: beryllium`

**Comment by jfindlay** [comment]
We need to make sure these file.replace changes make it into the beryllium release notes.

**Comment by lorengordon** [comment]
Ahh blast. I was updating my fork and I think I borked the PR. Should we close this one? I can resubmit it properly from a separate branch so it doesn't collide with other commits like that.

**Comment by jfindlay** [comment]
@lorengordon, sure, no problem.  Just be sure to reference this pull request in the new one.

**Pull request closed without merging**